src/tiledbsoma_ml/query_ids.py

```diff
@@ -0,0 +1,198 @@
+# Copyright (c) 2021-2024 The Chan Zuckerberg Initiative Foundation
+# Copyright (c) 2021-2024 TileDB, Inc.
+#
+# Licensed under the MIT License.
+"""Shuffle-chunk and partition (across GPU and DataLoader-worker processes) while reading from a SOMA |Experiment|."""
+from __future__ import annotations
+
+import logging
+from typing import (
+    List,
+    Literal,
+    Optional,
+    Tuple,
+)
+
+import numpy as np
+from attrs import define, evolve
+from tiledbsoma import (
+    ExperimentAxisQuery,
+)
+
+from tiledbsoma_ml._distributed import get_distributed_world_rank, get_worker_world_rank
+from tiledbsoma_ml._utils import batched, splits
+from tiledbsoma_ml.common import NDArrayJoinId
+
+logger = logging.getLogger("tiledbsoma_ml.query_ids")
+
+
+Chunks = List[NDArrayJoinId]
+r"""Return-type of |QueryIDs.shuffle_chunks|, |List| of |ndarray|\ s."""
+SamplingMethod = Literal["deterministic", "multinomial", "stochastic_rounding"]
+r"""Enum arg to |QueryIDs.split|:
+
+- ``"deterministic"``: number of each class returned will always be :math:`frac \times N`, rounded to nearest int, e.g.
+  ``n=12, fracs=[.7,.3]`` will always produce 8 and 4 elements, resp.
+- ``"multinomial"``: each element is assigned to a class independently; no guarantees are made about resulting class
+  sizes.
+- ``"stochastic_rounding"``: guarantee each class gets assigned at least :math:`\lfloor frac \times N \rfloor` elements.
+  The remainder are then distributed so that class-size expected-values match the provided ``fracs``.
+"""
+SamplingMethods = ["deterministic", "multinomial", "stochastic_rounding"]
```

Comment:
why is this needed?  I.e., why not just use `typing.get_args(SamplingMethod)`?

Reply:
How would you feel about an `Enum`?

Reply:
I think literal is the right solution in Python, and `typing.get_args` makes it DRY.  Not sure Enum adds anything.

that said, I'll let Ryan pick!